aof: change flag from a bool to a path

cb22 opened the pull request:
This allows specifying the absolute path of an AOF, and perhaps most interestingly allow for the AOF to be on a different drive to the data file.

~This should be called out in the release notes explicitly - while `--aof` is behind `--experimental`, this is a breaking CLI change.~

The change is fully backwards compatible.

cb22 committed:
aof: change flag from a bool to a path

This allows specifying the absolute path of an AOF, and perhaps most
interestingly allow for the AOF to be on a different drive to the data
file.

This should be called out in the release notes explicitly - while `--aof`
is behind `--experimental`, this is a breaking CLI change.

matklad commented on src/tigerbeetle/main.zig:
```diff
@@ -345,18 +345,18 @@ const Command = struct {
-            std.log.info("{s}", .{aof_path});
+        var aof: ?AOF = if (args.aof) |aof_path| blk: {
+            if (!std.mem.endsWith(u8, aof_path, ".aof")) {
+                vsr.fatal(.cli, "AOF path must end with .aof: '{s}'", .{aof_path});
```
Push this up to cli.zig, and instead assert here. As much as possible, we want to do all cli validation in cli.zig, and use assertions to double-check at the time of use. Actually, the assertion can be pushed down all the way to AOF.init?

matklad commented on src/tigerbeetle/main.zig:
```diff
@@ -345,18 +345,18 @@ const Command = struct {
+
+            const aof_dir = std.fs.path.dirname(aof_path) orelse ".";
+            const aof_dir_fd = try IO.open_dir(aof_dir);
+            errdefer std.posix.close(aof_dir_fd);
```
Needs to be `defer`, otherwise it'll leak

cb22 replied:
Ah, yeah, I still had the old code in my head which needed dir_fd to be kept around.

matklad commented:
We don't support AOF with auto-upgrades, right? So we shouldn't worry about version skew here?

matklad reviewed (approved):
Approving, assuming there's no upgrade interactions here

cb22 commented:
> We don't support AOF with auto-upgrades, right? So we shouldn't worry about version skew here?

Yeah, `--aof` requires `--experimental` which disables upgrade polling. You have to explicitly restart TigerBeetle in those cases, but that said it's worse than I thought initially, thinking about it again:

* stop all incoming traffic
* restart into new binary without `--aof` at all, let upgrade finish
* restart again with `--aof=<path>`
* resume incoming traffic

matklad commented:
This is no worse than before, no? You'd still have to do the dance of aof/no-aof/aof again?

cb22 commented:
> This is no worse than before, no? You'd still have to do the dance of aof/no-aof/aof again?

We changed it so that upgrade polling is disabled with `--experimental` but upgrades themselves still happen, so you could just do a single restart. But that's a good point, it was this way originally before that.

matklad commented:
I guess we can have `--aof` and `--aof-file` and have `--aof` desugare to `--aof-file`?

cb22 commented:
Yeah, I like the idea of having mutually exclusive flags better than having a single flag and another one to control it (eg, --aof and --aof-path, vs --aof or --aof-file).

There is another mildly cursed workaround that would allow us to avoid adding a flag: set `--aof=true` explicitly, then handle the string case "true" specially. Since we require .aof suffix there can't be a mix-up...

matklad commented:
>There is another mildly cursed workaround that would allow us to avoid adding a flag: set --aof=true explicitly, then handle the string case "true" specially. Since we require .aof suffix there can't be a mix-up...

I think you'll need to support both `--aof` and `--aof=true` in the same binary, which I'd rather avoid.

matklad reviewed (changes requested):
Retrackting approval just in case, I _think_ we are still not sure how to handle upgrades here!

cb22 commented:
> I think you'll need to support both --aof and --aof=true in the same binary, which I'd rather avoid.

(we have supported flags for boolean arguments since 7b35af2ec1522c53a1f600aa7adc453f63288ca9 FWIW, but I'm not sure that this goes back far enough to make the complexity of the workaround worth it. I'm going to add `--aof-file` which supersedes `--aof` but keep both as valid.)

cb22 committed:
aof: keep --aof and add in mutually exclusive --aof-file

cb22 commented on src/tigerbeetle/cli.zig:
```diff
@@ -915,7 +922,7 @@ fn parse_args_start(start: CLIArgs.Start) Command.Start {
         @intCast(@divExact(lsm_manifest_memory, constants.lsm_manifest_node_size));
 
-    return .{
+    var command_start: Command.Start = .{
```
Open to better ways of managing the buffer for generated aof path for `--aof`!

cb22 replied:
(I figured we'd want to keep all backwards compatibility CLI logic inside cli.zig...?)

matklad replied:
I think this won't work, the Start structure can move, invalidating interior pointer. You'll have to use a helper function here:

```zig
pub const Command = union(enum) {
    pub const Start = struct {
        aof_file: ?[:0]const u8,
        aof_file_buffer: [std.fs.max_path_bytes]u8 = .{0} ** std.fs.max_path_bytes,
    
        pub fn get_aof_path(start: *const Start) ?[:0]const u8 {
            if (start.aof_file) |file| return file;
            if (start.aof_file_buffer[0] != 0) {
                const index = std.mem.indexOfScalar(u8, start.aof_file_buffer, 0).?;
                assert(index > 0);
                return &start.aof_file_buffer[0..index :0];
            }
        }
    };

```

cb22 replied:
> the Start structure can move, invalidating interior pointer

True :confused:. I'm not a fan of adding a helper function and relying on the null termination of the string either however. How about a ~owned string type of `stdx.BoundedArrayType(u8, std.fs.max_path_bytes)`?

matklad replied:
That works as well! Why do we need `:0` anyway? Can we just remove nultermination from everywhere?

cb22 replied:
Almost everywhere yes, and done in [bae3791](https://github.com/tigerbeetle/tigerbeetle/pull/3055/commits/bae3791b120084f4bd9480a701ed62eb1c8b2de6)!

cb22 committed:
cli: remove null termination in most places

It's only needed for multiversioning's `exe_path`, since that's passed
through to various io_uring calls which expect the null termination.

matklad commented:
```
+- zig build-exe tigerbeetle Debug x86_64-linux 1 errors
zig/lib/std/meta.zig:916:12: error: evaluation exceeded 2000 backwards branches
    inline for (fields(T), 0..) |field, i| {
    ~~~~~~~^~~
zig/lib/std/meta.zig:916:12: note: use @setEvalBranchQuota() to raise the branch limit from 2000
src/tigerbeetle/cli.zig:758:39: note: called from here
            assert(std.meta.fieldIndex(@TypeOf(start), stable_arg) != null);
                   ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
referenced by:
    parse_args: src/tigerbeetle/cli.zig:634:55
```

cb22 committed:
cli: bump setEvalBranchQuota
